number_theory/modular.h

```diff
@@ -111,6 +112,120 @@ class Modular {
   }
 };
 
+namespace modular_internal {
+
+// Tests if the type T is the Modular class.
+template <typename T>
+struct IsModularImpl {
+  template <auto mod,
+            std::enable_if_t<std::is_same_v<T, Modular<mod>>, bool> = true>
+  static constexpr std::true_type test(Modular<mod>);
+
+  static constexpr std::false_type test(...);
+
+  static constexpr bool result = decltype(test(std::declval<T>()))::value;
+};
+
+}  // namespace modular_internal
+
+template <typename T>
+concept IsModular = modular_internal::IsModularImpl<T>::result;
+
+// Overloads an arithmetic operator with a class method
+#define OVERLOAD_ARITHMETIC_OPERATOR(CONCEPT, OP, METHOD) \
+  template <CONCEPT M>                                    \
```

Comment:
We can avoid concepts here like
```cpp
template <auto mod>
Modular<mod> operator OP(const Modular<mod> &lhs, const Modular<mod> &rhs)
{
  // ...
}
```

But, this restricts the type to be Modular and prevents further use of the macro to overload operators for other types, such as PrimeField.

Reply:
I think we got some miscommunication here due to the fact that I forgot our class name is `Modular`. When I was mentioning `Modular`, I was basically referring to `IsModular` concept since I already suggested renaming it to `Modular`. However, the name conflicts with the class name so we should rename it to something like `ModularType` I guess.